File: python/src/attesta/webhooks.py
"""Webhook notifications for attesta pipeline events.

Built on top of the :mod:`attesta.events` system, :class:`WebhookDispatcher`
subscribes to configured event types and sends HTTP POST requests with
JSON payloads to external endpoints.

Usage::

    from attesta.events import EventBus, EventType
    from attesta.webhooks import WebhookConfig, WebhookDispatcher

    bus = EventBus()
    config = WebhookConfig(
        url="https://hooks.example.com/attesta",
        events=[EventType.APPROVED, EventType.DENIED],
        secret="my-shared-secret",
    )
    dispatcher = WebhookDispatcher(bus, [config])

Zero external dependencies -- uses :mod:`urllib.request` from stdlib.
"""

from __future__ import annotations

import hashlib
import hmac
import json
import logging
import threading
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any
from urllib.error import URLError
from urllib.request import Request, urlopen

from attesta.events import Event, EventBus, EventType

__all__ = ["WebhookConfig", "WebhookDispatcher"]

logger = logging.getLogger("attesta.webhooks")


@dataclass
class WebhookConfig:
    """Configuration for a single webhook endpoint.

    Parameters
    ----------
    url:
        HTTP(S) endpoint to POST events to.
    events:
        List of event types to send. An empty list means *all* events.
    secret:
        Optional shared secret for HMAC-SHA256 signature verification.
        When set, an ``X-Attesta-Signature`` header is included.
    timeout:
        HTTP request timeout in seconds.
    retry_count:
        Number of retries on failure (0 = no retries).
    """

    url: str
    events: list[EventType] = field(default_factory=list)
    secret: str | None = None
    timeout: float = 5.0
    retry_count: int = 1


class WebhookDispatcher:
    """Subscribes to an EventBus and dispatches HTTP webhooks.

    Webhook deliveries happen in background threads (fire-and-forget)
    so they never block the approval pipeline.

    Parameters
    ----------
    event_bus:
        The :class:`EventBus` to subscribe to.
    configs:
        List of :class:`WebhookConfig` defining where to send events.
    """

    def __init__(
        self, event_bus: EventBus, configs: list[WebhookConfig]
    ) -> None:
        self._configs = configs
        self._event_bus = event_bus

        # Subscribe to all relevant event types.
        subscribed: set[EventType] = set()
        for config in configs:
            types = config.events or list(EventType)
            for event_type in types:
                if event_type not in subscribed:
                    event_bus.on(event_type, self._handle_event)
                    subscribed.add(event_type)

    def _handle_event(self, event: Event) -> None:
        """Route an event to the appropriate webhook configs."""
        for config in self._configs:
            if config.events and event.type not in config.events:
                continue
            # Fire in a background thread to avoid blocking the pipeline.
            thread = threading.Thread(
                target=self._send,
                args=(config, event),
                daemon=True,
            )
            thread.start()

    def _send(self, config: WebhookConfig, event: Event) -> None:
        """Send the webhook HTTP request (with retries)."""
        payload = self._build_payload(event)
        body = json.dumps(payload, separators=(",", ":")).encode("utf-8")

        headers: dict[str, str] = {
            "Content-Type": "application/json",
        }
        if config.secret:
            sig = hmac.new(
                config.secret.encode("utf-8"),
                body,
                hashlib.sha256,
            ).hexdigest()
            headers["X-Attesta-Signature"] = f"sha256={sig}"

        attempts = 1 + config.retry_count
        for attempt in range(attempts):
            try:
                req = Request(
                    config.url,
                    data=body,
                    headers=headers,
                    method="POST",
                )
                with urlopen(req, timeout=config.timeout) as resp:
                    status = resp.status
                    if 200 <= status < 300:
                        logger.debug(
                            "Webhook delivered to %s (status=%d)",
                            config.url,
                            status,
                        )
                        return
                    logger.warning(
                        "Webhook to %s returned status %d (attempt %d/%d)",
                        config.url,
                        status,
                        attempt + 1,
                        attempts,
                    )
            except (URLError, OSError, TimeoutError) as exc:
                logger.warning(
                    "Webhook to %s failed (attempt %d/%d): %s",
                    config.url,
                    attempt + 1,
                    attempts,
                    exc,
                )

            # Brief delay before retry.
            if attempt < attempts - 1:
                time.sleep(0.5)

        logger.error(
            "Webhook to %s exhausted all %d attempts",
            config.url,
            attempts,
        )

    @staticmethod
    def _build_payload(event: Event) -> dict[str, Any]:
        """Build the JSON payload for a webhook delivery."""
        return {
            "event": event.type.value.upper(),
            "timestamp": datetime.fromtimestamp(
                event.timestamp, tz=timezone.utc
            ).isoformat(),
            "data": event.data,
        }

File: typescript/src/webhooks.ts
/**
 * Webhook notifications for attesta pipeline events.
 *
 * Built on top of the events system, WebhookDispatcher subscribes to
 * configured event types and sends HTTP POST requests with JSON payloads
 * to external endpoints.
 *
 * Uses the global `fetch` API (available in Node.js 18+).
 */

import { createHmac } from "node:crypto";
import type { Event, EventBus, EventType } from "./events.js";

// ---------------------------------------------------------------------------
// WebhookConfig
// ---------------------------------------------------------------------------

/**
 * Configuration for a single webhook endpoint.
 */
export interface WebhookConfig {
  /** HTTP(S) endpoint to POST events to. */
  url: string;

  /** Event types to send. Empty array means all events. */
  events?: EventType[];

  /** Optional shared secret for HMAC-SHA256 signature. */
  secret?: string;

  /** HTTP request timeout in milliseconds. Default 5000. */
  timeoutMs?: number;

  /** Number of retries on failure (0 = no retries). Default 1. */
  retryCount?: number;
}

// ---------------------------------------------------------------------------
// WebhookDispatcher
// ---------------------------------------------------------------------------

/**
 * Subscribes to an EventBus and dispatches HTTP webhooks.
 *
 * Webhook deliveries are fire-and-forget (async, non-blocking).
 */
export class WebhookDispatcher {
  private readonly _configs: WebhookConfig[];

  constructor(eventBus: EventBus, configs: WebhookConfig[]) {
    this._configs = configs;

    // Subscribe to all relevant event types.
    const subscribed = new Set<string>();
    for (const config of configs) {
      const types = config.events ?? [];
      if (types.length === 0) {
        // Subscribe to all by using a catch-all
        if (!subscribed.has("*")) {
          // We need to import EventType values at subscription time
          const allTypes = [
            "risk_scored",
            "trust_computed",
            "challenge_issued",
            "challenge_completed",
            "approved",
            "denied",
            "audit_logged",
          ];
          for (const t of allTypes) {
            if (!subscribed.has(t)) {
              eventBus.on(
                t as EventType,
                (event: Event) => this._handleEvent(event)
              );
              subscribed.add(t);
            }
          }
          subscribed.add("*");
        }
      } else {
        for (const eventType of types) {
          if (!subscribed.has(eventType)) {
            eventBus.on(eventType, (event: Event) =>
              this._handleEvent(event)
            );
            subscribed.add(eventType);
          }
        }
      }
    }
  }

  private _handleEvent(event: Event): void {
    for (const config of this._configs) {
      if (
        config.events &&
        config.events.length > 0 &&
        !config.events.includes(event.type)
      ) {
        continue;
      }
      // Fire-and-forget
      this._send(config, event).catch((err) => {
        // eslint-disable-next-line no-console
        console.error(`[attesta] Webhook to ${config.url} failed:`, err);
      });
    }
  }

  private async _send(config: WebhookConfig, event: Event): Promise<void> {
    const payload = buildPayload(event);
    const body = JSON.stringify(payload);

    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };

    if (config.secret) {
      const sig = createHmac("sha256", config.secret)
        .update(body)
        .digest("hex");
      headers["X-Attesta-Signature"] = `sha256=${sig}`;
    }

    const attempts = 1 + (config.retryCount ?? 1);
    const timeoutMs = config.timeoutMs ?? 5000;

    for (let attempt = 0; attempt < attempts; attempt++) {
      try {
        const controller = new AbortController();
        const timer = setTimeout(() => controller.abort(), timeoutMs);

        const response = await fetch(config.url, {
          method: "POST",
          headers,
          body,
          signal: controller.signal,
        });

        clearTimeout(timer);

        if (response.ok) return;

        // eslint-disable-next-line no-console
        console.warn(
          `[attesta] Webhook to ${config.url} returned ${response.status} (attempt ${attempt + 1}/${attempts})`
        );
      } catch (err) {
        // eslint-disable-next-line no-console
        console.warn(
          `[attesta] Webhook to ${config.url} failed (attempt ${attempt + 1}/${attempts}):`,
          err
        );
      }

      // Brief delay before retry
      if (attempt < attempts - 1) {
        await new Promise((resolve) => setTimeout(resolve, 500));
      }
    }
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function buildPayload(event: Event): Record<string, unknown> {
  return {
    event: event.type.toUpperCase(),
    timestamp: new Date(event.timestamp).toISOString(),
    data: event.data,
  };
}
